[iOS] [UnifiedPDF] Support text interactions and selection in rotated PDF pages

whsieh opened the pull request:
#### 0a6c80bc21cba9b3b65f6d7eec39d711f970295d
<pre>
[iOS] [UnifiedPDF] Support text interactions and selection in rotated PDF pages
<a href="https://bugs.webkit.org/show_bug.cgi?id=286599">https://bugs.webkit.org/show_bug.cgi?id=286599</a>
<a href="https://rdar.apple.com/141946781">rdar://141946781</a>

Reviewed by Abrar Rahman Protyasha.

Add support for proper text selection rendering and text interactons in rotated pages of a PDF
document when unified PDF is enabled. See below for more details.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::selectionCaretPointInPage const):

Rewrite this helper method to account for axis-aligned rotations (i.e. 90°, 180°, or 270°) when
finding the locations of the start and end character positions in the selected range. Currently,
this method always assumes that the end of a selected piece of text is close to the right edge of
the selection rect, while the start is close to the left edge; however, this strategy only works
when the PDF is not rotated at all.

To make this more robust, we map selection rects for the given `PDFSelection` into the root view,
and then choose the midpoint of the top, left, bottom or right edge of the root view rect depending
on which way the PDF is rotated. We determine this by calculating the `cumulativeTransform` when
mapping from PDF page to content coordinates, and then extracting just the approximate rotation in
degrees (rounding to the nearest 90°) from this transform using the `a` and `b` values of the affine
transformation.

Finally, once we&apos;ve determined which edge is closest to the start or end position of the selection,
we then remap this back to PDF page coordinates.

(WebKit::UnifiedPDFPlugin::platformPopulateEditorStateIfNeeded const):

Honor transforms on each of the selection rects&apos; bounds when collecting selection geometries for
`EditorState` in PDFs.

(WebKit::UnifiedPDFPlugin::selectionCaretPointInPage): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Add a new test PDF that&apos;s rotated 90 degrees clockwise.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Add a new API test to exercise this change, by sanity checking the dimensions of a piece of selected
text selection in a rotated PDF.

Canonical link: <a href="https://commits.webkit.org/289481@main">https://commits.webkit.org/289481@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f4a846ba83c4993310d5ef4fa2551ad773e6515

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6625 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41474 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91974 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/37854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89165 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14693 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67323 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/37854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90117 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5275 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78847 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47645 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5050 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33223 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36971 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/75541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34102 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93861 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14277 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/76128 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14481 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74703 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75327 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18536 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19670 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/7194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14296 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14041 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17484 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15822 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -4241,13 +4241,48 @@ static bool isEmpty(PDFSelection *selection)
+    AffineTransform cumulativeTransform = [page transformForBox:kPDFDisplayBoxMediaBox];
+    bool appliedLineTransform = false;
+    [selectedLine enumerateRectsAndTransformsForPage:page.get() usingBlock:[&](CGRect rect, CGAffineTransform transform) {
+        if (std::exchange(appliedLineTransform, true))
```
So we apply the transform for the first rect only?

whsieh replied:
Ah, so in theory, there should only ever be a single rect (since this is enumerating rects along just the first line). I can add some assertions to help clarify (and enforce) this.

aprotyas replied:
Oh, right, that makes sense.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -4241,13 +4241,48 @@ static bool isEmpty(PDFSelection *selection)
-    return { WTFMove(page), { offsetX, lineBounds.y() + (lineBounds.height() / 2) } };
+    // FIXME: Account for RTL text and vertical writing mode.
+    return rootViewToPage([&] -> FloatPoint {
+        int clockwiseRotationAngle = static_cast<int>(360 + 90 * std::round(-rad2deg(rotationInRadians) / 90)) % 360;
```
![image](https://github.com/user-attachments/assets/3784b052-cdd6-456b-bad7-8a78aa0ae0c1)

whsieh replied:
😰

aprotyas reviewed (approved):
This is great!

whsieh committed:
[iOS] [UnifiedPDF] Support text interactions and selection in rotated PDF pages
https://bugs.webkit.org/show_bug.cgi?id=286599
rdar://141946781

Reviewed by Abrar Rahman Protyasha.

Add support for proper text selection rendering and text interactons in rotated pages of a PDF
document when unified PDF is enabled. See below for more details.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::selectionCaretPointInPage const):

Rewrite this helper method to account for axis-aligned rotations (i.e. 90°, 180°, or 270°) when
finding the locations of the start and end character positions in the selected range. Currently,
this method always assumes that the end of a selected piece of text is close to the right edge of
the selection rect, while the start is close to the left edge; however, this strategy only works
when the PDF is not rotated at all.

To make this more robust, we map selection rects for the given `PDFSelection` into the root view,
and then choose the midpoint of the top, left, bottom or right edge of the root view rect depending
on which way the PDF is rotated. We determine this by calculating the `cumulativeTransform` when
mapping from PDF page to content coordinates, and then extracting just the approximate rotation in
degrees (rounding to the nearest 90°) from this transform using the `a` and `b` values of the affine
transformation.

Finally, once we've determined which edge is closest to the start or end position of the selection,
we then remap this back to PDF page coordinates.

(WebKit::UnifiedPDFPlugin::platformPopulateEditorStateIfNeeded const):

Honor transforms on each of the selection rects' bounds when collecting selection geometries for
`EditorState` in PDFs.

(WebKit::UnifiedPDFPlugin::selectionCaretPointInPage): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Add a new test PDF that's rotated 90 degrees clockwise.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Add a new API test to exercise this change, by sanity checking the dimensions of a piece of selected
text selection in a rotated PDF.

Canonical link: https://commits.webkit.org/289481@main
